src/pipewire/pipewire.cc

```diff
@@ -159,36 +161,61 @@ void PipeWireOutput::pause(bool pause)
 {
     pw_thread_loop_lock(m_loop);
     pw_stream_set_active(m_stream, !pause);
+    clock_gettime(CLOCK_REALTIME, &m_time);
     pw_thread_loop_unlock(m_loop);
 }
 
 int PipeWireOutput::get_delay()
 {
-    return (m_buffer_at / m_stride + m_frames) * 1000 / m_rate;
+    int buff_time = (((m_buffer.len() + m_buffer_size) / m_stride) * 1000) / m_rate;
+
+    // Get time difference from last buffer fill
+    timespec time_now {};
+    clock_gettime(CLOCK_REALTIME, &time_now);
+
+    int time_diff =
+        (((time_now.tv_sec - m_time.tv_sec) * 1000000000) +
+         (time_now.tv_nsec - m_time.tv_nsec)) / 1000000;
+
+    if (time_diff < 0)
+        time_diff = 0;
+    if (time_diff > buff_time)
+        time_diff = buff_time;
+
+    return buff_time - time_diff;
 }
 
 void PipeWireOutput::drain()
 {
     pw_thread_loop_lock(m_loop);
-    if (m_buffer_at > 0)
-        pw_thread_loop_timed_wait(m_loop, 2);
+
+    int buflen;
+    while ((buflen = m_buffer.len()) > 0)
+    {
+        pw_thread_loop_timed_wait(m_loop, 1);
+        if (buflen <= m_buffer.len())
+        {
+            AUDERR("PipeWireOutput: buffer drain lock\n");
+            break;
+        }
+    }
 
     pw_stream_flush(m_stream, true);
-    pw_thread_loop_timed_wait(m_loop, 2);
+    pw_thread_loop_timed_wait(m_loop, 1);
```

Comment:
What is this extra `timed_wait` for? Seems like we could use an explanatory comment here.

Reply:
pw_thread_loop_timed_wait(m_loop, 1) is required after pw_stream_flush(m_stream, true);
otherwise on_drained(void * data) is not triggered.
Will be comment in code on that.